resource-recorder/src/dal.rs

```diff
@@ -0,0 +1,413 @@
+
+        let pool = SqlitePool::connect_with(sqlite_options).await.unwrap();
+
+        Self::from_pool(pool).await
+    }
+
+    #[allow(dead_code)]
+    pub async fn new_in_memory() -> Self {
+        let pool = SqlitePool::connect("sqlite::memory:").await.unwrap();
+        Self::from_pool(pool).await
+    }
+
+    async fn from_pool(pool: SqlitePool) -> Self {
+        MIGRATIONS.run(&pool).await.unwrap();
+
+        Self { pool }
+    }
+}
+
+#[async_trait]
+impl Dal for Sqlite {
+    async fn add_resources(
+        &self,
+        project_id: Ulid,
+        service_id: Ulid,
+        resources: Vec<Resource>,
+    ) -> Result<(), sqlx::Error> {
+        let mut transaction = self.pool.begin().await?;
+
+        sqlx::query("UPDATE resources SET is_active = false WHERE service_id = ?")
+            .bind(service_id.to_string())
+            .execute(&mut transaction)
+            .await?;
+
+        // Making mutliple DB "connections" is fine since the sqlite is on the same machine
+        for resource in resources {
+            if let Some(r_project_id) = resource.project_id {
+                if r_project_id != project_id {
+                    warn!("adding a resource that belongs to another project");
+                }
+            }
+
+            if let Some(r_service_id) = resource.service_id {
+                if r_service_id != service_id {
+                    warn!("adding a resource that belongs to another service");
+                }
+            }
+
+            sqlx::query("INSERT OR REPLACE INTO resources (project_id, service_id, type, config, data, is_active) VALUES(?, ?, ?, ?, ?, ?)")
+            .bind(project_id.to_string())
+            .bind(service_id.to_string())
+            .bind(resource.r#type)
+            .bind(resource.config)
+            .bind(resource.data)
+            .bind(resource.is_active)
+            .execute(&mut transaction)
+            .await?;
+        }
+
+        transaction.commit().await
+    }
+
+    async fn get_project_resources(&self, project_id: Ulid) -> Result<Vec<Resource>, sqlx::Error> {
+        sqlx::query_as(r#"SELECT * FROM resources WHERE project_id = ?"#)
+            .bind(project_id.to_string())
+            .fetch_all(&self.pool)
+            .await
+    }
+
+    async fn get_service_resources(&self, service_id: Ulid) -> Result<Vec<Resource>, sqlx::Error> {
+        sqlx::query_as(r#"SELECT * FROM resources WHERE service_id = ?"#)
+            .bind(service_id.to_string())
+            .fetch_all(&self.pool)
+            .await
+    }
+
+    async fn delete_resource(&self, resource: &Resource) -> Result<(), sqlx::Error> {
+        sqlx::query("DELETE FROM resources WHERE project_id = ? AND service_id = ? AND type = ?")
+            .bind(resource.project_id.map(|u| u.to_string()))
+            .bind(resource.service_id.map(|u| u.to_string()))
+            .bind(resource.r#type)
+            .execute(&self.pool)
+            .await
+            .map(|_| ())
```

Comment:
I'm not sure yet. So far I'm thinking one of two routes:
1. Something else (deployer) call this service and provisioner. And it will check if a resource is inactive before trying to delete it.
2. This service will be the only service that calls provisioner. Then we should check in `lib.rs` if the resource is inactive first.

I'm not sure which route we will take yet. We'll see when we stick all the services together. So I'm leaving it for now.